lib/io.dx

```diff
@@ -0,0 +1,77 @@
+
+'File system operations
+
+FilePath : Type = String
+data CString = MkCString CharPtr
+
+data StreamMode =
+  ReadMode
+  WriteMode
+
+data Stream mode:StreamMode = MkStream CharPtr
+
+-- TODO: check the string contains no nulls
+def withCString (s:String) (action: CString -> {State World} a) : {State World} a =
+  (AsList n s') = s <> (AsList _ "\NUL")
+  withTabPtr s' \(MkPtr ptr). action $ MkCString ptr
+
+def fopen (path:String) (mode:StreamMode) : {State World} (Stream mode) =
+  modeStr = AsList _ case mode of
+    ReadMode  -> "r"
+    WriteMode -> "w"
+  withCString path \(MkCString pathPtr).
+    withCString modeStr \(MkCString modePtr).
+      MkStream $ %ffi fopen CharPtr pathPtr modePtr
+
+def fclose (mode:StreamMode) ?-> (stream:Stream mode) : {State World} Unit =
```

Comment:
Would it be reasonable to make `fopen` and `fclosed` not public so people have to use `withFile`, and thus we can be sure everything that is opened is closed?

Reply:
To me, whether to support APIs (only "bracketed" safe operations) is a decision that should be driven by use cases, it is hard to judge independently at face value: are there use cases for supporting unsafe `fopen` and `fclose` operations? Not supporting unsafe operations to start would be safest, but supporting them doesn't hurt much - we can advocate for users to use `withFile`.

---

@apaszke recently ran into trouble with `llvm-hs` where [the library provided only bracketed operations](https://github.com/llvm-hs/llvm-hs/blob/a48170078dcf492860a00e70dedb65833f6aef54/llvm-hs/src/LLVM/Internal/OrcJIT.hs#L65-L79):

```haskell
-- | Specifies how external symbols in a module added to a
-- 'CompileLayer' should be resolved.
newtype SymbolResolver =
  SymbolResolver (MangledSymbol -> IO (Either JITSymbolError JITSymbol))

-- | Create a `FFI.SymbolResolver` that can be used with the JIT.
withSymbolResolver :: ExecutionSession -> SymbolResolver -> (Ptr FFI.SymbolResolver -> IO a) -> IO a
withSymbolResolver (ExecutionSession es) (SymbolResolver resolverFn) f = ...
```

This required him to create additional custom shims:

https://github.com/google-research/dex-lang/blob/fc825c119d7a5cc45c1abd0263cfe2234390161e/src/lib/LLVM/Shims.hs#L33-L47